Use Celluloid, Reel and Celluloid I/O based http requests

azul opened the pull request:
Http gem + celluloid have a much simpler API that can do without the blocks we use all over the place. This pull request does not make use of this yet. It's just putting them in place of the old event machine based things.
Afterwards nickserver will need a solid cleanup. But for now this should give us proper ssl validation :).

azul committed:
refactor: change send_response -> respond in em_server

This way the interface mimics the one of Celluloid Reels request.

request.respond status, content

azul committed:
first attempt at using celluloid instead of em - broken

Main problem right now is the outdated version of webmock.

azul committed:
update webmock version

azul committed:
Trying to replace EM base server with reel

some tests are still broken. But at least they are running now.

azul committed:
fix host header handling and tests

azul committed:
actually use celluloid io for http requests

azul committed:
use proper ssl connection with ca_file - green!

elijh commented:
it is not required to use the fiber-based api of celluloid, so what is a big refactor needed? i agree that it is much nicer, but now that a reactor based code is working, why bother to rewrite it?

elijh commented on Gemfile:
```diff
@@ -2,3 +2,6 @@ source 'https://rubygems.org'
 # Specify your gem's dependencies in keymachine.gemspec
 gemspec
+
+gem 'byebug'
```
why not `add_development_dependency('byebug')` in gemspec?

elijh commented:
looks good!

elijh commented:
maybe also remove this from the README:

> Uses asynchronous network IO for both server and client connections (via EventMachine)

azul committed:
minor: silence warnings and cleanup

azul committed:
refactor: replace blocks/yields with returns

This became possible because we now use celluloid.
Celluloid handles asynchronity without the need for callbacks
or blocks.

azul committed:
refactor: turn EmailAddress into a class, cleanup

azul commented:
@elijh I was not aware there is a non fiber api of celluloid. We are already using the fiber one here - but we use it and then yield the result as we did in callbacks before. I changed this in a new commit. I'll show you what i mean there: https://github.com/leapcode/nickserver/pull/5/commits/fb2d7e6f8f1fceefbc8964d34369a867eb8f25bb

azul commented on lib/nickserver/adapters/celluloid_http.rb:
```diff
@@ -13,7 +13,7 @@ def get(url, options = {})
         ssl_context: ctx,
         ssl_socket_class: Celluloid::IO::SSLSocket
-      yield response.code, response.to_s
+      return response.code, response.to_s
```
@elijh this is what i meant above. We were already using fiber based API - but instead of returning the results we kept the blocks on our side. This resulted in quite a bit of handing down blocks. I think our code is much cleaner in this pseudo synchronous style.

azul commented on lib/nickserver/hkp/source.rb:
```diff
@@ -14,22 +14,20 @@
-        else
-          yield Nickserver::Response.new(status, response)
-        end
+    def query(nick)
```
This is one example of simplification by moving from blocks to return values.

azul commented on lib/nickserver/hkp/source.rb:
```diff
@@ -13,22 +14,20 @@
+      status, response = search(nick)
+      if status == 200
+        best = pick_best_key(response)
+        get_key_by_fingerprint(nick, best.keyid)
```
no more handing down the &block

azul commented on lib/nickserver/hkp/source.rb:
```diff
@@ -13,22 +14,20 @@
+        best = pick_best_key(response)
+        get_key_by_fingerprint(nick, best.keyid)
+      else
+        Nickserver::Response.new(status, response)
```
This is one example of simplification by moving from blocks to return values.
no more yielding results and no more handing down the &block in line 21

azul committed:
fix start / stop script

I have to admit I don't completely understand what is going on here.
https://github.com/celluloid/celluloid/wiki/Frequently-Asked-Questions says
you can either use sleep after initiating the actors or call run on them.

But this might not be true for reel. At least run did not work and now
new; sleep seems to work fine.
